### PR TITLE
Retrieve Context from a ContentProvider

### PR DESCRIPTION
**Pull request by SalvatoreT**
Remove the need to have the context passed-in on `Picasso.with()`.

https://github.com/square/picasso/issues/1634

### CODE REVIEW

**Comment by SalvatoreT** [comment]
@JakeWharton, here's a first swing at it.

**Comment by NightlyNexus** [comment]
Not everybody wants a ContentProvider or static instance of a Picasso in his or her app.

File: picasso/src/main/java/com/squareup/picasso/Picasso.java [hunk 1]
```diff
@@ -674,16 +673,9 @@ void shutdown() {
...
-  public static Picasso with(@NonNull Context context) {
-    if (context == null) {
-      throw new IllegalArgumentException("context == null");
-    }
+  public static Picasso with() {
     if (singleton == null) {
-      synchronized (Picasso.class) {
-        if (singleton == null) {
-          singleton = new Builder(context).build();
-        }
```
Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/Picasso.java]: We should still do the initialization here. Just stash the Context in a static field on this class from the content provider.

File: picasso/src/main/java/com/squareup/picasso/PicassoProvider.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+package com.squareup.picasso;
+
+import android.content.ContentProvider;
+import android.content.ContentValues;
+import android.database.Cursor;
+import android.net.Uri;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+public class PicassoProvider extends ContentProvider {
```
Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/PicassoProvider.java]: nit: final

Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/PicassoProvider.java]: Also can you annotate this with `@RestrictTo(LIBRARY)` as an extra measure.

File: picasso/src/main/java/com/squareup/picasso/Picasso.java (hunk 1 above)
Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/Picasso.java]: Actually stash the Context inside a static on the content provider itself. That will prevent this class from initializing until it's referenced.

File: picasso/src/main/java/com/squareup/picasso/PicassoProvider.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  @Override public boolean onCreate() {
+    Picasso.setSingletonInstance(new Picasso.Builder(getContext()).build());
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/PicassoProvider.java]: if you want to override the singleton instance, then, you need to remove the ContentProvider from the manifest, right?
or is there another way?

**Comment by SalvatoreT** [comment]
@JakeWharton & @NightlyNexus, ready for a second pass.

- [ ] squash commits

File: picasso/src/main/java/com/squareup/picasso/Picasso.java [hunk 4]
```diff
@@ -689,6 +688,16 @@ public static Picasso with(@NonNull Context context) {
   }
 
   /**
+   * The global default {@link Picasso} instance.
+   * <p>
+   * This instance is created by passing a global context from startup into {@link #with(Context)}.
+   * </p>
+   */
+  public static Picasso with() {
+    return with(PicassoProvider.context);
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/Picasso.java]: nice, still lazy and still lets setSingletonInstance work fine without disabling anything.

**Comment by pyricau** [comment]
Suggestion: should the name be something else than `with()` ? "with context" makes sense, "with" by itself is a bit weird.

**Comment by JakeWharton** [comment]
This was discussed in the follow-up where I removed the content version
completely.

**Pull request merged**